Add penalty-based mortar equal value constraint

recuero opened the pull request:
Refs. #21329

recuero committed:
Add penalty-based mortar equal value constraint (#21329)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1082303/) on 1c5c446 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21338/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1082491/) on ac07289 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>74571c</th>
      <th colspan=4>#21338 ac0728</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.46%</td>
      <td>83.46%</td>
      <td>+0.00%</td>
      <td>88.89%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77431</td>
      <td>77447</td>
      <td>+16</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15342</td>
      <td>15344</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21338/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21338/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21338/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 88.89% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on test/tests/mortar/continuity-3d-non-conforming/continuity_penalty_sphere_hex20.i:
```diff
@@ -0,0 +1,126 @@
+  []
+[]
+
+[Problem]
+  kernel_coverage_check = false
+[]
```
Is this necessary? (I can believe that it is ... I would hope it's unnecessary in the case where you have an LM, but I could believe that we get an error if the LM doesn't exist)

recuero replied:
I get a catastrophic segmentation fault at the very beginning of the run if I don't include that for the penalty tests.

lindsayad replied:
catastrophic segmentation fault? That sounds even worse than a normal segmentation fault 😄 We should probably make an issue for that

lindsayad replied:
opened #21416

lindsayad commented on test/tests/mortar/continuity-3d-non-conforming/continuity_penalty_sphere_hex20.i:
```diff
@@ -0,0 +1,126 @@
+[Executioner]
+  solve_type = NEWTON
+  type = Steady
+  petsc_options_iname = '-pc_type -snes_linesearch_type -pc_factor_shift_type '
+                        '-pc_factor_shift_amount'
+  petsc_options_value = 'lu       basic                 NONZERO               1e-15'
```
Can you do `boomeramg` with this problem? I'm curious since we no longer have a saddle point

recuero replied:
Yeah, we can do BoomerAMG with the penalty tests. Using the same BoomerAMG options for the LM-based enforcement doesn't make these small tests converge. So, I guess things are making sense.

lindsayad reviewed (commented):
Impressive number of tests!

recuero committed:
Address Alex's review

The penalty approach to mortar equality constraint allows these tests to be
solved with the Boomer algebraic multigrid method preconditioner.

recuero commented:
Moving from `superlu` to `boomerAMG` caused the penalty hex20 test to time out. Made it `heavy`. It appears it's the only heavy test in the framework.

lindsayad commented:
How long did it take with LU? If it's like a 2X increase or more, then I'd vote to go back to LU

recuero commented:
With `opt`, locally, it goes from 8.16s (LU) to 11.8 s (BoomerAMG). Since on CIVET it's compiled with `O0`, this difference puts it above 300s.

lindsayad commented:
That's a pretty insane differential. From 11.8 seconds to > 300 going from `-O2` to `-O0` ... yikes. I don't suppose there's a way to make the mesh smaller?

recuero committed:
Make hex20 test with coarser mesh
